tests/test_dataclasses.py

```diff
@@ -1395,17 +1395,54 @@ class Foo:
         Foo(**{'x': '1', 'y': '2'})
 
 
-@pytest.mark.xfail(reason='recursive references need rebuilding?')
+@pytest.mark.xfail(reason='need to make it possible to rebuild dataclasses')
 def test_self_reference_dataclass():
     @pydantic.dataclasses.dataclass
     class MyDataclass:
-        self_reference: 'MyDataclass'
+        self_reference: Optional['MyDataclass'] = None
 
-    annotation = MyDataclass.__pydantic_fields__['self_reference'].annotation
-    # Currently, this is true:
-    # assert isinstance(MyDataclass, PydanticForwardRef)
-    # TODO: Probably need a way to "model_rebuild" a dataclass
-    assert annotation is MyDataclass
+    # rebuild_pydantic_dataclass(MyDataclass)
+
+    assert MyDataclass.__pydantic_fields__['self_reference'].annotation == Optional[MyDataclass]
+
+    instance = MyDataclass(self_reference=MyDataclass(self_reference=MyDataclass()))
+    assert AnalyzedType(MyDataclass).dump_python(instance) == {
+        'self_reference': {'self_reference': {'self_reference': None}}
+    }
+
+    with pytest.raises(ValidationError) as exc_info:
+        MyDataclass(self_reference=MyDataclass(self_reference=1))
+
+    assert exc_info.value.errors() == [
+        {
+            'ctx': {'dataclass_name': 'MyDataclass'},
+            'input': 1,
+            'loc': ('self_reference',),
+            'msg': 'Input should be a dictionary or an instance of MyDataclass',
+            'type': 'dataclass_type',
+        }
+    ]
+
+
+@pytest.mark.xfail(reason='need to make it possible to rebuild dataclasses')
+def test_cyclic_reference_dataclass():
+    @pydantic.dataclasses.dataclass
+    class D1:
+        d2: Optional['D2'] = None
+
+    @pydantic.dataclasses.dataclass
+    class D2:
+        d1: Optional[D1] = None
+
+    # rebuild_pydantic_dataclass(D1)
+
+    instance = D1(d2=D2(d1=D1(d2=D2(d1=D1()))))
+
+    assert AnalyzedType(D1).dump_python(instance) == {...}
```

Comment:
Why are we using AnalyzedType in these tests? Are we actually testing it or does it make it easier to test the other stuff?

Reply:
Parametrized generic dataclasses don't do validation based on the parameters. (This is related to _GenericAlias calling methods on the origin class without the ability to reflect the arguments.) So the only way you can get validation for generic dataclasses that reflects the parameters is to have them be a field of something (a BaseModel, pydantic dataclass, or TypedDict), or to use AnalyzedType. AnalyzedType seemed to best-convey the intent here. Probably deserves a comment explaining this.

Reply:
Oh makes sense, thank you for explaining :)